### ISSUE TITLE
pdf not rendering well with Bootstrap 5.1

### ISSUE THREAD

**Issue by ccerrillo** [dompdf/dompdf#2828]
Here is a sample: https://eclecticgeek.com/dompdf/debug.php?identifier=b0db706513e58e10b57a30f7b8805a61
It's rendering well in source HTML tab but in the pdf table rows are written over each other and the h1 is not displayed

```
<!DOCTYPE html>
<html lang="es">
  <head>
    <meta charset="utf-8">
    <title>Test</title>
    <link href="https://cdn.jsdelivr.net/npm/bootstrap@5.1.0/dist/css/bootstrap.css" rel="stylesheet">
  </head>
  <body>
  <h1>Informe</h1>
  <table class="table">
    <thead>
      <tr>
        <th scope="col">ID</th>
        <th scope="col">Nombre</th>
        <th scope="col">Email</th>
        <th scope="col">Total invertido</th>
        <th scope="col">Beneficio</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <th scope="row">1</th>
          <td>Alvaro</td>
          <td>al@va.ro</td>
          <td>20.00</td>
          <td>0.50</td>
        </tr>
    </tbody>
  </table>
  </body>
</html>
```

**Comment by bsweeney** [comment]
My general answer is that Dompdf does not support Bootstrap at this time (#631). But specifically what appears to be the problem here is that Bootstrap 5 relies heavily on CSS variables, which are not handled at all (#1872). Additionally, some properties are being overwritten with invalid values due to an issue with how Dompdf currently parses styles (#2824).

You can sort of work around the issue by by overriding the style property values Dompdf is unable to handle. For example, add the following after the reference to Bootstrap to fix the overlap:

```
<style>
@media dompdf {
   * { line-height: 1.2; }
}
</style>
```

The dompdf media type limits the style impact to Dompdf.

Of course, depending on the complexity of your document this may not be a feasible way to address the issue since you'd have to override a significant number of declarations.

**Comment by Rpgdudester** [comment]
I would like to mention that domPDF does work surprisingly well with bootstrap 3.3.7. I'm running bootstrap 5.1 and I couldn't get it to work well with dompdf, however, I was able to load 3.3.7 (make sure it's in your system and not loading from a url and use 

'<style>' . file_get_contents(css/bootstrap.min.css") . '</style>' to load it into your css. 

just make sure you use col-xs-* for the columns and you may still have to do some padding/margin modifications to make it work within . 

This is for people who still want to still want to use bootstrap to make pdf files. 

Thanks Brian for an awesome software!

**Comment by solmazb128** [comment]
but i have to use bootstrap to have a faithfull pdf from the content fetched from ckeditor.
if dompdf could not include bootstrap properly, is there any other alternative ?

**Comment by bsweeney** [comment]
Dompdf can render Bootstrap to some degree, and my earlier comment about variable support was addressed with the 3.0.0 release. However, Dompdf does not support all features of Bootstrap so the issue primarily depends on your use case.

With Bootstrap 5 lack of support for flexbox is probably going to be an issue. And generally there's the question around resolution. PDFs have a low resolution so depending on the Dompdf DPI and page size you might render under a SM or XS Bootstrap breakpoint. And one of the problematic lacking features is box-sizing, which can be worked around using a custom stylesheet but would be tedious to build out.
